Add System Overview Widget to System Map view

Web-eWorks opened the pull request:
This PR adds the System Overview widget (the body list in the top right of the world view) to both of the System Map view modes. It also refactors out the layout code from both the Sector Map and System Map into a single reusable module for creating full-screen UIs.

It also changes the focus-object interaction in the SystemView to use double-clicks instead of single-clicks; this may be a slightly controversial change, but having selection state decoupled from focus state is fairly useful when interacting with the overview widget.

This is likely the last of the work branches I have outstanding to be merged into the release, nothing else is close enough to done to be a viable candidate.

impaktor commented:
> It also refactors out the layout code from both the Sector Map and System Map into a single reusable module for creating full-screen UIs.

Perchance, a base could be painted on this.... onto which one could walk.... and stabby-stabby-stabby people, and take their loot, and in a bar listen to their tales of great adventure, and stuff. (One of these days, I'll muster up the energy to get cracking on this, (as the British say)).

Web-eWorks commented:
> Perchance, a base could be painted on this.... onto which one could walk.... and stabby-stabby-stabby people, and take their loot, and in a bar listen to their tales of great adventure, and stuff. (One of these days, I'll muster up the energy to get cracking on this, (as the British say)).

Unfortunately probably not, this is specifically designed for creating a UI on top of other fullscreen interactable content (e.g. Flight UI, Map screens, etc.). For something like the Bar screen I might have a few ideas up my sleeve in the future to make it a bit easier to build, but I can't really say anything concrete at this juncture. Having a proper mockup / concept art really helps make UI a reality though, so perhaps @nozmajner would be happy to help... in the appropriate discussion venue, of course.

Gliese852 commented:
when I try to open a system map, for example, in Barnard Star, I get an assertion
`src/galaxy/StarSystem.cpp:202: SystemBody* StarSystem::GetBodyByPath(const SystemPath&) const : Assertion 'm_path.IsSameSystem(path)' failed.`

Gliese852 commented:
How can I hide system-overview-window in the system map?

Gliese852 commented:
'center' context menu item does not center

Web-eWorks commented:
> when I try to open a system map, for example, in Barnard Star, I get an assertion `src/galaxy/StarSystem.cpp:202: SystemBody* StarSystem::GetBodyByPath(const SystemPath&) const : Assertion 'm_path.IsSameSystem(path)' failed.`

Dangit, that's what I get for running without assertions enabled. Will look into that.

The system overview widget does not have a hide button in the map screen, although that is probably something worth adding.

Web-eWorks commented on data/pigui/libs/window-layout.lua:
```diff
@@ -0,0 +1,106 @@
+
+local layout = utils.inherits(nil, 'ui.Layout')
+
+local __newLayout = layout.New
```
My personal stylistic convention for "local variable cache"; e.g. when overriding a global function you'd write `local __print = print; print = function(...) __print(...) end`. Doesn't have any special significance, just a convention I use occasionally.

Web-eWorks commented on data/pigui/modules/flight-ui/system-overview.lua:
```diff
@@ -0,0 +1,85 @@
+
+	self:drawControlButtons()
+
+	ui.sameLine(ui.getWindowSize().x - (button_size.x + 10))
```
Probably a stand-in for `itemSpacing.x`. I wrote this several months ago... I'll clean it up and replace it with a properly scaled value.

Web-eWorks commented on data/pigui/modules/flight-ui/system-overview.lua:
```diff
@@ -0,0 +1,85 @@
+
+local icons = ui.theme.icons
+
+local button_size = Vector2(32,32) * (ui.screenHeight / 1200)
```
Yes, this was a straight port of the existing code without too much effort to clean it all up and make it pretty. Will do a cleanup pass.

Web-eWorks commented on data/pigui/modules/system-overview-window.lua:
```diff
@@ -1,197 +1,227 @@
+	result = make_result(systemBody, systemBody.name, isSelected)
+	result.visible = isSelected or filter(systemBody)
+	if systemBody.isSpaceStation then
+		parent.has_space_stations = true
```
Wow, using my own advice against me! :stuck_out_tongue: Good catch, though that behavior is integral to the algorithm it's not documented.

Web-eWorks replied:
After some consideration, this particular case is one where I'd consider the code to be "self-documenting"; it's not part of any API accessible outside of the module and the mutated argument is only ever used in the scope of the function itself via recursion. I've added a line comment clarifying the scope and purpose of the mutation however for maximum clarity.

Also, if I went around adding explicit comments to every function in Pioneer that modifies a table it's passed I'd be here for a very long time... the original advice was made in the context of value-like objects (e.g. vectors) where the conceptual model is pass-by-value but the underlying object model does not allow such. Tables are firmly pass-by-reference, and while it's "nice" to document the fact that a function modifies a table it received from outside execution context, functions that mutate tables to create data structures are a well-defined idiom at this point.

Web-eWorks commented on data/pigui/libs/wrappers.lua:
```diff
@@ -507,7 +507,7 @@ function ui.tabBarFont(id, items, font, ...)
 		end
 
-		pigui.EndTabItem()
+		pigui.EndTabBar()
```
I had not woken up nearly as much as I thought when I authored this commit, good catch. This isn't even supposed to be on this branch, I thought I was pushing to `microfixes`.

Web-eWorks commented on data/pigui/views/map-sector-view.lua:
```diff
@@ -432,72 +424,50 @@ end
-local dummyFrames = 3
+function sectorViewLayout:onUpdateWindowConstraints(w)
+	-- resizing, aligning windows - static
+	w.current.pos = edgePadding
```
The vector-copy expression isn't very ergonomic, so I'd rather just guarantee that layout functions don't modify the vector value. If it is being modified in a non-trivial manner then I'll write the extra verbosity needed for copying.

Gliese852 commented on data/pigui/modules/system-overview-window.lua:
```diff
@@ -1,197 +1,232 @@
+end
+
+function SystemOverviewWidget:display(system, root, selected)
+	self:overrideDrawButtons()
```
why not call `drawControlButtons()` right away?

impaktor commented:
Ah, I get it now. The new System Atlas Mode (#5239) only shows orbital bodies, not stations on them (like the old version did), so this will solve that problem neatly.

Gliese852 commented:
> when I try to open a system map, for example, in Barnard Star, I get an assertion
> `src/galaxy/StarSystem.cpp:202: SystemBody* StarSystem::GetBodyByPath(const SystemPath&) const : Assertion 'm_path.IsSameSystem(path)' failed.`

@Web-eWorks here is lua traceback:
```
[C](-1): ?                                                                                                                                                                                                                                                                       
[C](-1): __index                                                                                                                                                                                                                                                                 
[string "[T] @pigui/modules/system-overview-window.lua..."](77): calculateEntry                                                                                                                                                                                                  
[string "[T] @pigui/modules/system-overview-window.lua..."](224): fun                                                                                                                                                                                                            
[string "[T] @pigui/libs/wrappers.lua"](235): child                                                                                                                                                                                                                              
[string "[T] @pigui/modules/system-overview-window.lua..."](223): display                                                                                                                                                                                                        
[string "[T] @pigui/modules/system-view-ui.lua"](205): fun                                                                                                                                                                                                                       
[string "[T] @pigui/libs/wrappers.lua"](543): withFont                                                                                                                                                                                                                           
[string "[T] @pigui/modules/system-view-ui.lua"](204): Show                                                                                                                                                                                                                      
[string "[T] @pigui/libs/window-layout.lua"](74): fun                                                                                                                                                                                                                            
[string "[T] @pigui/libs/wrappers.lua"](90): window                                                                                                                                                                                                                              
[string "[T] @pigui/libs/window-layout.lua"](73): fun                                                                                                                                                                                                                            
[string "[T] @pigui/libs/wrappers.lua"](543): withFont                                                                                                                                                                                                                           
[string "[T] @pigui/libs/window-layout.lua"](69): display                                                                                                                                                                                                                        
[string "[T] @pigui/modules/system-view-ui.lua"](718): window_fnc                                                                                                                                                                                                                
[string "[T] @pigui/baseui.lua"](49): fun                                                                                                                                                                                                                                        
[string "[T] @pigui/libs/wrappers.lua"](90): window                                                                                                                                                                                                                              
[string "[T] @pigui/baseui.lua"](47): ?
```

as I understand it, this is due to the fact that you use the `body` attribute on the systemBody, and for remote systems, the systemBody does not have a real body. I counted ~10 accesses to this attribute in `system-overview-window.lua` and all that will not work when the system map does not display the current system.

Web-eWorks commented:
> as I understand it, this is due to the fact that you use the body attribute on the systemBody, and for remote systems, the systemBody does not have a real body. I counted ~10 accesses to this attribute in system-overview-window.lua and all that will not work when the system map does not display the current system.

Fixed. That assert is something that Lua shouldn't be able to trigger (and doesn't have any easy way to tell if a system body has a real body associated without querying that variable) so I added some error handling for that case.
